File: backend/routers/albums.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from backend import database, models, auth
from pydantic import BaseModel

router = APIRouter(
    prefix="/albums",
    tags=["albums"],
)

class AlbumUpdate(BaseModel):
    new_name: Optional[str] = None
    artist: Optional[str] = None
    genre: Optional[str] = None

class AlbumPublish(BaseModel):
    is_public: bool

@router.get("/{album_name}", response_model=dict)
def get_album_details(
    album_name: str,
    current_user: Optional[models.User] = Depends(auth.get_current_user_optional),
    db: Session = Depends(database.get_db)
):
    # Decode URL-encoded album name if necessary (FastAPI usually handles this, but good to be safe with spaces)
    # Actually, let's trust FastAPI's handling but keep in mind frontend sends encoded strings
    
    # Strategy: Find tracks with this album name.
    # We need to be careful: multiple users might have an album named "Greatest Hits".
    # For now, since album names aren't unique per user in our model (just a string on Track),
    # we might get collisions if we don't scope by user. 
    # BUT, the request implies viewing a specific album.
    # To fix the collision issue properly, the URL should probably include the uploader_id, e.g. /albums/{user_id}/{album_name}
    # However, the current requirements didn't specify that complexity. 
    # Let's try to infer intent:
    # If looking up by name, we might find multiple.
    # Implementation Plan adjustment: Let's assume for this iteration we query purely by name, 
    # but practically we should filter.
    # Refined logic:
    # If the user clicks from Profile, we know the User ID.
    # If we just use album_name, it's ambiguous.
    # LET'S ADD QUERY PARAM `uploader_id`.
    
    query = db.query(models.Track).filter(models.Track.album == album_name)
    
    # If uploader_id is passed in query (optional but recommended for uniqueness)
    # We don't have it in the route path defined in plan, but we can accept it as query param
    # For now, let's just get all tracks with that album name (simplistic approach).
    # Wait, if I upload "Hits" and you upload "Hits", they mix? Yes, with current model.
    # To mitigate, I will enforce that we group by uploader if possible, but the UI needs to trigger this.
    # Let's stick to the plan but maybe filter if `current_user` matches?
    # No, that logic is flawed if viewing others.
    
    # FIX: We will accept `uploader_id` as a query parameter optionally.
    # If not provided, it fetches specific album tracks? No, that's messy.
    # Step-back: `get_user_albums` in `users.py` groups by uploader.
    # So when we click an album, it's usually associated with a user.
    # I'll add `uploader_id` as a required query param to be safe/correct?
    # Or just fetch all and separate? 
    # Let's try to infer from the requested tracks.
    
    tracks = query.all()
    
    if not tracks:
        raise HTTPException(status_code=404, detail="Album not found")

    # Filter logic similar to user tracks
    visible_tracks = []
    
    # We need to detect if we are viewing "mixed" albums (same name, diff users) or just one.
    # Ideally, frontend passes uploader_id.
    # I'll assume for this milestone we treat strict string matching.
    
    for track in tracks:
        if track.is_public:
            visible_tracks.append(track)
        elif current_user and track.uploader_id == current_user.id:
            visible_tracks.append(track)
            
    if not visible_tracks and not (current_user and any(t.uploader_id == current_user.id for t in tracks)):
        # If no visible tracks and user isn't an owner of any, returns 404 effectively
        raise HTTPException(status_code=404, detail="Album not found or private")

    # If simple string match gives us multiple users' albums, checking "owner" is tricky.
    # Let's Pick the "primary" uploader (e.g. the first one found or majority) -- insecure.
    # Better: Filter by the uploader of the first track found? 
    # No, let's return all matching the name that are visible.
    
    # Prepare response
    # Metadata from the first track
    first_track = visible_tracks[0] if visible_tracks else tracks[0] # Fallback if owner but all private (shouldn't happen with logic above)
    
    # Determine if "Editor" (Owner)
    # User is editor if they own ALL the tracks? Or ANY?
    # Let's say: User is owner if they own the displayed tracks.
    is_owner = False
    if current_user:
        if all(t.uploader_id == current_user.id for t in visible_tracks):
            is_owner = True

    # Get user favorites
    user_favorites = set()
    if current_user:
        favs = db.query(models.Favorite).filter(models.Favorite.user_id == current_user.id).all()
        user_favorites = {f.track_id for f in favs}

    response = {
        "name": first_track.album,
        "artist": first_track.artist, # Approximate
        "album_art_path": first_track.album_art_path, # Take first available
        "total_duration": sum(t.duration for t in visible_tracks),
        "track_count": len(visible_tracks),
        "is_owner": is_owner,
        "tracks": [
            {
                "id": t.id,
                "title": t.title,
                "artist": t.artist,
                "duration": t.duration,
                "is_public": t.is_public,
                "uploader_id": t.uploader_id,
                "is_favorite": t.id in user_favorites
            } for t in visible_tracks
        ]
    }
    return response

@router.put("/{album_name}")
def update_album(
    album_name: str,
    update: AlbumUpdate,
    current_user: models.User = Depends(auth.get_current_user),
    db: Session = Depends(database.get_db)
):
    # Find tracks with this album name belonging to CURRENT USER ONLY
    tracks = db.query(models.Track).filter(
        models.Track.album == album_name,
        models.Track.uploader_id == current_user.id
    ).all()
    
    if not tracks:
        raise HTTPException(status_code=404, detail="Album not found in your library")
        
    for track in tracks:
        if update.new_name:
            track.album = update.new_name
        if update.artist:
            track.artist = update.artist
        if update.genre:
            track.genre = update.genre
            
    db.commit()
    return {"message": f"Updated {len(tracks)} tracks", "new_name": update.new_name or album_name}

@router.put("/{album_name}/publish")
def publish_album(
    album_name: str,
    publish_data: AlbumPublish,
    current_user: models.User = Depends(auth.get_current_user),
    db: Session = Depends(database.get_db)
):
    tracks = db.query(models.Track).filter(
        models.Track.album == album_name,
        models.Track.uploader_id == current_user.id
    ).all()
    
    if not tracks:
        raise HTTPException(status_code=404, detail="Album not found in your library")
        
    for track in tracks:
        track.is_public = publish_data.is_public
        
    db.commit()
    return {"message": f"Album {'published' if publish_data.is_public else 'unpublished'} ({len(tracks)} tracks updated)"}

@router.delete("/{album_name}")
def delete_album(
    album_name: str,
    current_user: models.User = Depends(auth.get_current_user),
    db: Session = Depends(database.get_db)
):
    import os
    tracks = db.query(models.Track).filter(
        models.Track.album == album_name,
        models.Track.uploader_id == current_user.id
    ).all()
    
    if not tracks:
        raise HTTPException(status_code=404, detail="Album not found in your library")
        
    deleted_count = 0
    for track in tracks:
        # Delete file
        if os.path.exists(track.file_path):
            try:
                os.remove(track.file_path)
            except Exception as e:
                print(f"Error deleting file {track.file_path}: {e}")
        
        db.delete(track)
        deleted_count += 1
        
    db.commit()
    return {"message": f"Deleted album '{album_name}' and {deleted_count} tracks"}

File: backend-node/routes/albums.js
const express = require('express');
const router = express.Router();
const path = require('path');
const fs = require('fs');
const db = require('../db');
const { requireAuth, optionalAuth } = require('../auth');

router.get('/:id', optionalAuth, (req, res) => {
    const albumId = req.params.id;

    const album = db.prepare(`
        SELECT a.*, ar.name as artist 
        FROM albums a
        LEFT JOIN artists ar ON a.artist_id = ar.id
        WHERE a.id = ?
    `).get(albumId);

    if (!album) return res.status(404).json({ detail: 'Album not found' });

    const tracks = db.prepare(`
        SELECT t.*, a.album_art_path, a.title as album, ar.name as artist 
        FROM tracks t
        LEFT JOIN albums a ON t.album_id = a.id
        LEFT JOIN artists ar ON t.artist_id = ar.id
        WHERE t.album_id = ?
    `).all(album.id);

    const visibleTracks = tracks.filter(track => {
        if (track.is_public) return true;
        if (req.user && track.uploader_id === req.user.id) return true;
        return false;
    });

    if (visibleTracks.length === 0 && tracks.length > 0) {
        return res.status(404).json({ detail: 'Album tracks are private' });
    }

    const favoriteIds = req.user
        ? new Set(db.prepare('SELECT track_id FROM favorites WHERE user_id = ?').all(req.user.id).map(f => f.track_id))
        : new Set();

    const isOwner = req.user && album.uploader_id === req.user.id;

    res.json({
        id: album.id,
        name: album.title,
        artist: album.artist,
        artist_id: album.artist_id,
        album_art_path: album.album_art_path,
        total_duration: visibleTracks.reduce((sum, t) => sum + (t.duration || 0), 0),
        track_count: visibleTracks.length,
        is_owner: isOwner,
        tracks: visibleTracks.map(t => ({
            id: t.id,
            title: t.title,
            artist: t.artist,
            artist_id: t.artist_id,
            album_id: t.album_id,
            album: t.album,
            duration: t.duration,
            is_public: !!t.is_public,
            uploader_id: t.uploader_id,
            album_art_path: t.album_art_path,
            is_favorite: favoriteIds.has(t.id)
        }))
    });
});

router.put('/:id', requireAuth, (req, res) => {
    const albumId = req.params.id;
    const { new_name, artist, genre } = req.body;

    const album = db.prepare('SELECT * FROM albums WHERE id = ?').get(albumId);
    if (!album) return res.status(404).json({ detail: 'Album not found' });
    if (album.uploader_id !== req.user.id) return res.status(403).json({ detail: 'Not authorized' });

    if (new_name) {
        db.prepare('UPDATE albums SET title = ? WHERE id = ?').run(new_name, album.id);
    }
    
    if (artist) {
        let artistId;
        const existingArtist = db.prepare('SELECT id FROM artists WHERE name = ?').get(artist);
        if (existingArtist) {
            artistId = existingArtist.id;
        } else {
            const res = db.prepare('INSERT INTO artists (name) VALUES (?)').run(artist);
            artistId = res.lastInsertRowid;
        }

        db.prepare('UPDATE albums SET artist_id = ? WHERE id = ?').run(artistId, album.id);
        db.prepare('UPDATE tracks SET artist_id = ? WHERE album_id = ?').run(artistId, album.id);
    }
    
    if (genre) {
        db.prepare('UPDATE tracks SET genre = ? WHERE album_id = ?').run(genre, album.id);
    }

    res.json({ message: `Updated album`, new_name: new_name || album.title, id: album.id });
});

router.put('/:id/publish', requireAuth, (req, res) => {
    const albumId = req.params.id;
    const { is_public } = req.body;

    const album = db.prepare('SELECT * FROM albums WHERE id = ?').get(albumId);
    if (!album) return res.status(404).json({ detail: 'Album not found' });
    if (album.uploader_id !== req.user.id) return res.status(403).json({ detail: 'Not authorized' });

    const publicValue = is_public ? 1 : 0;
    db.prepare('UPDATE tracks SET is_public = ? WHERE album_id = ?').run(publicValue, album.id);

    res.json({ message: `Album ${is_public ? 'published' : 'unpublished'}` });
});

router.delete('/:id', requireAuth, (req, res) => {
    const albumId = req.params.id;
    const album = db.prepare('SELECT * FROM albums WHERE id = ?').get(albumId);
    if (!album) return res.status(404).json({ detail: 'Album not found' });
    if (album.uploader_id !== req.user.id) return res.status(403).json({ detail: 'Not authorized' });

    const tracks = db.prepare('SELECT * FROM tracks WHERE album_id = ?').all(album.id);
    let deletedCount = 0;
    for (const track of tracks) {
        db.prepare('DELETE FROM playlist_tracks WHERE track_id = ?').run(track.id);
        db.prepare('DELETE FROM favorites WHERE track_id = ?').run(track.id);
        db.prepare('DELETE FROM tracks WHERE id = ?').run(track.id);

        const trackPath = path.join(__dirname, '..', '..', track.file_path);
        if (fs.existsSync(trackPath)) {
            try { fs.unlinkSync(trackPath); } catch (e) { console.error(`Error deleting file`, e); }
        }
        deletedCount++;
    }

    if (album.album_art_path) {
        const artPath = path.join(__dirname, '..', '..', album.album_art_path);
        if (fs.existsSync(artPath)) {
             try { fs.unlinkSync(artPath); } catch (e) { console.error(`Error deleting art`, e); }
        }
    }

    db.prepare('DELETE FROM albums WHERE id = ?').run(album.id);
    res.json({ message: `Deleted album '${album.title}' and ${deletedCount} tracks` });
});

module.exports = router;